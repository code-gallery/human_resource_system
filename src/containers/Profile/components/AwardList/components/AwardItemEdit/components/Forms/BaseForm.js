import React from 'react'
import PropTypes from 'prop-types'
import { organizationsAutocomplete } from 'utils/networkRequests'

class BaseForm extends React.Component {
  handleChange = (value, fieldName) => {
    this.props.onFieldChange(fieldName, value)
  }

  handleMultiChange = fields => {
    this.props.onFieldsChange(fields)
  }

  onTextChange = (fieldName, event) => {
    if (fieldName === 'location' && event && event.formatted_address) {
      this.props.onFieldChange('location', event.formatted_address)
    } else {
      this.props.onFieldChange(fieldName, event.target.value)
    }
  }

  getOptions = (q) => {
    return organizationsAutocomplete(q).then(res => {
      return { options: res }
    })
  }

  optionRenderer(option) {
    return (
      <div className="institution-select">
        <img src={option.logo_image ? option.logo_image : '/images/appii-placeholder.png'} alt={option.name} />
        <p>{option.name}</p>
        <span>{option.country}</span>
      </div>
    )
  }

  selectAsyncChange = (v) => {
    this.handleMultiChange({
      organisation: v ? v.name : '',
      organisation_id: v ? v.id : null
    })
    this.props.orgRegistered(true)
  }

  selectAsyncBlur = (event) => {
    if (event && event.target.value && event.target.value.length) {
      this.handleMultiChange({
        organisation: event.target.value,
        organisation_id: null
      })
      this.props.orgRegistered(false)
    } else {
      this.handleChange(this.props.organisation, 'organisation')
    }
  }

  render() {
    return null
  }
}

BaseForm.defaultProps = {
  orgRegistered: function() {}
}

BaseForm.propTypes = {
  organisation: PropTypes.string,
  onFieldChange: PropTypes.func.isRequired,
  onFieldsChange: PropTypes.func.isRequired,
  orgRegistered: PropTypes.func
}

export default BaseForm
