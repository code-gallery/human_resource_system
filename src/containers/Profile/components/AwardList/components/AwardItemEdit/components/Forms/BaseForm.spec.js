import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import BaseForm from './BaseForm'

describe('containers/Profile/components/.../BaseForm', () => {
  const props = {
    onFieldChange: jest.fn(),
    onFieldsChange: jest.fn(),
    orgRegistered: jest.fn(),
    organisation: 'Applied'
  }

  let wrapper

  beforeEach(() => {
    wrapper = shallow(
      <BaseForm {...props} />
    )
    props.onFieldChange.mockClear()
    props.onFieldsChange.mockClear()
  })

  describe('handleChange', () => {
    it('calls props.onFieldChange correctly', () => {
      wrapper.instance().handleChange('Informal', 'cpd_type')
      expect(props.onFieldChange)
        .toHaveBeenCalledWith('cpd_type', 'Informal')
    })
  })

  describe('onTextChange', () => {
    describe('when field "location" changes', () => {
      it('calls props.onFieldChange correctly', () => {
        const event = { formatted_address: 'London, UK' }
        wrapper.instance().onTextChange('location', event)
        expect(props.onFieldChange)
          .toHaveBeenCalledWith('location', 'London, UK')
      })
    })

    describe('when field "test" changes', () => {
      it('calls props.onFieldChange correctly', () => {
        const event = { target: { value: 'hello world' } }
        wrapper.instance().onTextChange('test', event)
        expect(props.onFieldChange)
          .toHaveBeenCalledWith('test', 'hello world')
      })
    })
  })

  describe('optionRenderer', () => {
    describe('when logo is default', () => {
      it('renders correct element', () => {
        const option = {
          name: 'CPD1',
          country: 'UK'
        }

        const tree = renderer.create(
          wrapper.instance().optionRenderer(option)
        ).toJSON()

        expect(tree).toMatchSnapshot()
      })
    })

    describe('when logo exists', () => {
      it('renders correct element', () => {
        const option = {
          name: 'CPD1',
          country: 'UK',
          logo_image: 'https://appii.s3.eu-west-2.amazonaws.com/organisations/logo_image/12_logo.jpg'
        }

        const tree = renderer.create(
          wrapper.instance().optionRenderer(option)
        ).toJSON()

        expect(tree).toMatchSnapshot()
      })
    })
  })

  describe('selectAsyncChange', () => {
    it('calls props.onFieldChange correctly', () => {
      wrapper.instance().selectAsyncChange({ name: 'applied' })
      expect(props.onFieldsChange)
        .toHaveBeenCalledWith({ organisation: 'applied' })
    })

    it('calls props.orgRegistered correctly', () => {
      wrapper.instance().selectAsyncChange()
      expect(props.orgRegistered)
        .toHaveBeenCalledWith(true)
    })
  })

  describe('selectAsyncBlur', () => {
    describe('when event target is valid', () => {
      beforeEach(() => {
        wrapper.instance().selectAsyncBlur({
          target: { value: 'applied' }
        })
      })

      it('calls props.onFieldChange correctly', () => {
        expect(props.onFieldsChange)
          .toHaveBeenCalledWith({ organisation: 'applied', organisation_id: null })
      })

      it('calls props.orgRegistered correctly', () => {
        expect(props.orgRegistered)
          .toHaveBeenCalledWith(false)
      })
    })

    describe('when event is not defined', () => {
      it('calls props.onFieldChange correctly', () => {
        wrapper.instance().selectAsyncBlur()
        expect(props.onFieldChange)
          .toHaveBeenCalledWith('organisation', props.organisation)
      })
    })
  })

  describe('getOptions', () => {
    it('calls without error', () => {
      expect(function() {
        wrapper.instance().getOptions()
      }).not.toThrow()
    })
  })
})
