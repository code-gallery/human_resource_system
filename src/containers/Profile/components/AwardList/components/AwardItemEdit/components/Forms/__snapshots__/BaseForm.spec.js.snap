// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Profile/components/.../BaseForm optionRenderer when logo exists renders correct element 1`] = `
<div
  className="institution-select"
>
  <img
    alt="CPD1"
    src="https://appii.s3.eu-west-2.amazonaws.com/organisations/logo_image/12_logo.jpg"
  />
  <p>
    CPD1
  </p>
  <span>
    UK
  </span>
</div>
`;

exports[`containers/Profile/components/.../BaseForm optionRenderer when logo is default renders correct element 1`] = `
<div
  className="institution-select"
>
  <img
    alt="CPD1"
    src="/images/appii-placeholder.png"
  />
  <p>
    CPD1
  </p>
  <span>
    UK
  </span>
</div>
`;
